### DISCUSSION TITLE
Sane config??

### DISCUSSION THREAD

**Discussion by Hobbyist11** [fredrikaverpil/neotest-golang#160] [Configuration] [answered]
Hello, I'm quite new to neotest and neotest-golang, I have tried setting it up for a couple of days now however I'm not sure if I've set it up correctly as I'm facing the following issues: 1.Incredibly slow for a basic single file test (I believe this is caused by the tester seeing all the golang tests on another directory since when I open Neotest summary it goes up to the higher parent directories) after opening summary it also starts to make neovim incredibly laggy and slow.
2. It doesn't start testing a file  unless I save the file first and open summary otherwise it just says no test found. 

System details:
Artix Linux
Neovim v10.1
Plugins all on latest version

Things I've tried: Reinstallling go TServer, gomod tsserver, reinstalling my entire nvim config, trying out Nvim distros with Neotest already set-up and I get the same results. I wanted to try copying @fredrikaverpil 's config but couldn't get it to work,it couldn't find the neotest-golang install for some reason. Wanted to try neotest's minimal init too but I also couldn't get it to work since running that seems to ignore that the plugin is installed by lazy??

Thank you so much, would love to hear your thoughts and from your experiences!

My config is as follows:
Neotest:
```
local M = {
  "nvim-neotest/neotest",
  dependencies = {
    "nvim-neotest/nvim-nio",
    "antoinemadec/FixCursorHold.nvim",
    "nvim-treesitter/nvim-treesitter",
    -- general tests
    "vim-test/vim-test",
    "nvim-neotest/neotest-vim-test",
    -- language specific tests
    "fredrikaverpil/neotest-golang",
    "marilari88/neotest-vitest",
    "nvim-neotest/neotest-python",
    "nvim-neotest/neotest-plenary",
    "rouge8/neotest-rust",
    "lawrence-laz/neotest-zig",
    "rcasia/neotest-bash",
  },
}

function M.config()
  local wk = require "which-key"
  wk.add {
    {
      "<leader>tl",
      function()
        require("neotest").run.run_last()
      end,
      desc = "Run Last",
    },
    {
      "<leader>ts",
      function()
        require("neotest").summary.toggle()
      end,
      desc = "Toggle Summary",
    },
    { "<leader>tt", "<cmd>lua require'neotest'.run.run()<cr>", desc = "Test Nearest" },
    {
      "<leader>tT",
      function()
        require("neotest").run.run(vim.uv.cwd())
      end,
      desc = "Run All Test Files",
    },
    { "<leader>tf", "<cmd>lua require('neotest').run.run(vim.fn.expand('%'))<cr>", desc = "Test File" },
    { "<leader>td", "<cmd>lua require('neotest').run.run({strategy = 'dap'})<cr>", desc = "Debug Test" },
    -- { "<leader>ts", "<cmd>lua require('neotest').run.stop()<cr>", desc = "Test Stop" },
    { "<leader>ta", "<cmd>lua require('neotest').run.attach()<cr>", desc = "Attach Test" },
    -- {
    --   "<leader>tS",
    --   function()
    --     require("neotest").run.stop()
    --   end,
    --   desc = "Stop",
    -- },
    {
      "<leader>tw",
      function()
        require("neotest").watch.toggle(vim.fn.expand "%")
      end,
      desc = "Toggle Watch",
    },
    {
      "<leader>to",
      function()
        require("neotest").output.open { enter = true, auto_close = true }
      end,
      desc = "Show Output",
    },
    {
      "<leader>tO",
      function()
        require("neotest").output_panel.toggle()
      end,
      desc = "Toggle Output Panel",
    },
  }

  ---@diagnostic disable: missing-fields
  require("neotest").setup {
    adapters = {
      require "neotest-python" {
        dap = { justMyCode = false },
      },
      -- require "neotest-golang",
      require "neotest-golang",
      require "neotest-vitest",
      require "neotest-zig",
      require "neotest-vim-test" {
        ignore_file_types = { "python", "vim", "lua", "javascript", "typescript", "go" },
      },
    },
    status = { virtual_text = true },
    output = { open_on_run = true },
    log_level = vim.log.levels.DEBUG,
  }
end

return M
```

P.S looking at Neotest logs I see this (tried looking it up on the web but it seems I'm the  only one with this issue)
```
ERROR | 2024-08-05T20:46:21Z+0800 | ...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:309 | Couldn't find positions in path
lid subprocess call: Vim:Error invoking 'nvim_exec_lua' on channel 4:
Invalid channel: 4
stack traceback:
	[C]: in function 'assert'
	...l/share/nvim/lazy/neotest/lua/neotest/lib/subprocess.lua:147: in function 'call'
	...re/nvim/lazy/neotest/lua/neotest/lib/treesitter/init.lua:189: in function 'parse_positions'
	...re/nvim/lazy/neotest-golang/lua/neotest-golang/query.lua:138: in function 'discover_positions'
	...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:300: in function <...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:264>
	[C]: in function 'xpcall'
	...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:264: in function '_update_positions'
	...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:317: in function <...ocal/share/nvim/lazy/neotest/lua/neotest/client/init.lua:315>
```
Running :checkhealth neotest-golang, everything is fine, it even finds the proper go mod and in the right directory.

**Comment by fredrikaverpil** [comment]
Hey @Hobbyist11 did you check out the [tips & troubleshooting](https://github.com/fredrikaverpil/neotest-golang#%EF%B8%8F-tips--troubleshooting) section in the README?

It mentions logging (which you seem to be already doing) and also how you can speed up Neotest.

 1. Exactly _what_ is slow? Detection of the test or the execution? Does it matter if you disable all but the one adapter you were using?
2. You won't be able to run a test unless the file exists on disk, so this is expected. Or did you mean something else?

**Reply by fredrikaverpil** [reply]
> I wanted to try copying @fredrikaverpil 's config but couldn't get it to work,it couldn't find the neotest-golang install for some reason.

You need to remove the `dir` declaration from my config, as I am loading the adapter from a local development folder.

https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-fredrik/lua/lang/go.lua#L213

Could that have been it?

**Reply by fredrikaverpil** [reply]
One thing that struck me is that it looks like you are always loading all those adapters for various languages. This means that all of them are using tree-sitter to parse the files in your project for potential tests.

You can see how I'm only loading the neotest-golang adapter if there's a *.go file opened: https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-fredrik/lua/lang/go.lua#L209
I have other Neotest adapters in my own config, but they won't load if I'm not opening the relevant file types.

There could also be a problem with conflicting adapters that struggle to run in tandem. I would try to just enable one at a time and see at which point there are problems. For example, there was this (however, I believe the root cause was in Neotest itself for that particular issue): https://github.com/mrcjkb/rustaceanvim/issues/399

When looking at the error you provide, it seems truncated (see how line 2 is not a continuation of line 1), and it's difficult to tell by this error alone if this was triggered by one of those other adapters or what happened prior to this error.

**Reply by Hobbyist11** [reply]
> > I wanted to try copying @fredrikaverpil 's config but couldn't get it to work,it couldn't find the neotest-golang install for some reason.
> 
> You need to remove the `dir` declaration from my config, as I am loading the adapter from a local development folder.
> 
> https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-fredrik/lua/lang/go.lua#L213
> 
> Could that have been it?

Ah right, I didn't notice that, will try again! UPDATE: Using your config now and it seems to be working? I tried a test earlier and it worked fine, what I did was open test file, open summary then run the test. I closed nvim, open same test file and ran the test again without opening summary this time and it said, couldn't find test.

**Reply by fredrikaverpil** [reply]
That sounds odd. But it's difficult to help you here unless you can provide more information. 

Can you provide a minimal and reproducible test example?

**Reply by Hobbyist11** [reply]
> That sounds odd. But it's difficult to help you here unless you can provide more information.
> 
> Can you provide a minimal and reproducible test example?

Not sure if this would be a minimal and reproducible test but curious how this runs on your set up. Here's a video from mine and the code used:

So it basically can't find the test for some reason and it lags to the point that I just quit nvim.

https://github.com/user-attachments/assets/c4006d99-13e2-4d4d-a3be-dc0cd714aef4

Here on the second one I just use what I've been doing before which is running go test on the terminal which I find quite repetitive



https://github.com/user-attachments/assets/4d89e094-7b86-41e4-bd4f-85575890e04a

(My expectation was moving to neotest would allow me to skip the terminal and not have to type go test every time. I may be mistaken here? But am I correct to assume that with a proper working neotest-golang set up, it should just run the test and show me the result?)

The code:
go.mod
```
module main.go

go 1.22.5
```

main.go
```
package main

import "fmt"

func Hello() string {
	return "Hello, world"
}

func main() {
	fmt.Println(Hello())
}
```

hello_test.go
```
package main

import "testing"

func TestHello(t *testing.T) {
	got := Hello()
	want := "Hello, world"

	if got != want {
		t.Errorf("got %q want %q", got, want)
	}
}
```

**Reply by fredrikaverpil** [reply, marked as answer]
Your videos are flickering a lot here and are quite hard to follow when it comes exactly what you are doing. I'm also not sure what I'm looking at. It looks like you have cloned down this repo and then added files to the root of it which are then attempting to run?
Difficult to comment on.

However, the three files you provided I can work with. I can successfully run the test and there are no delays or waiting times in parsing tests etc. I'm not sure what can be wrong on your end and I also don't know how to help you here unless you provide more information.

https://github.com/user-attachments/assets/65fa129e-3038-4d21-86d8-0f88139a413f

**Reply by Hobbyist11** [reply]
> Your videos are flickering a lot here and are quite hard to follow when it comes exactly what you are doing. I'm also not sure what I'm looking at. It looks like you have cloned down this repo and then added files to the root of it which are then attempting to run? Difficult to comment on.
>>Ah yes, I didn't notice that I think it's a hardware issue (the flickering). Also, yes I basically have a folder of golang codes. Different directories are different tutorials I'm following though IDK why it goes up to the parent and reads the other dirs and not just the thing I'm currently working on which should be just GoWithTests but it also shows the other dir like Golangtuts and the Go installation folder itself..

> However, the three files you provided I can work with. I can successfully run the test and there are no delays or waiting times in parsing tests etc. I'm not sure what can be wrong on your end and I also don't know how to help you here unless you provide more information.
> bug3.mov
>>I guess my assumptions were correct. A properly working config really is what I expected. Idk why I can't duplicate it despite using the same config as yours :(


Also, thank you so much for taking your time to help me out on this, I really appreciate it! I think I've hit a roadblock now, I don't know what to do anymore. I guess I'll just try again after a while.

**Reply by Hobbyist11** [reply]
Oh one last thing, sorry but which version of nvim are running in the video you sent also would it work the same if you just opened nvim from home and navigated to the test file using neotree??
I'm running v0.10.1

**Reply by fredrikaverpil** [reply]
> which version of nvim are running in the video you sent

```bash
❯ nvim --version
NVIM v0.11.0-dev-3642+gb4b4cf46a-Homebrew
Build type: Release
LuaJIT 2.1.1720049189
Run "nvim -V1 -v" for more info
```

But you are good with 0.10.1 too. I ran that version a couple of days ago.

> would it work the same if you just opened nvim from home and navigated to the test file using neotree

You mean open nvim in the cloned down git repo folder `neotest-golang`?
That's what I'm doing all the time.

If you mean `~/` when you say "home folder", I wouldn't do that for various reasons (plugins would access your entire home folder, file/content searching with e.g. telescope would be very slow etc). But technically, yes, it would work. I just tried it here and I can still run the tests just fine when starting nvim in `~/` and opening `~/code/public/neotest-golang/tests/go/positions_test.go` but it was a bit yanky. If this is what you're doing, I would say this is very likely your main issue. A thumb of rule is to start nvim in your project folder. Then you can build a bunch of stuff on top of that assumption, like for example which tooling you might have available for that project (using e.g. pkgx) and which environment variables might be registered (using e.g. direnv).

**Reply by fredrikaverpil** [reply]
I see now, when going back to your videos that you 1) might be on Windows (your initial report mentioned "Artix Linux") and 2) you are launching nvim from the user home folder.

So; 

1. On Windows, you will most likely have to enable the yet undocumented gotestsum runner by providing the `runner = "gotestsum"` option when setting up neotest-golang: https://github.com/fredrikaverpil/neotest-golang/pull/150
2. Please start `nvim` from the Go project folder instead (regardless of OS). A lot of issues you might have here (slowness being one) would most likely stem from this problem. Imagine the sheer number of files that Neotest might be parsing that resides under your user home folder.

**Reply by Hobbyist11** [reply]
> I see now, when going back to your videos that you are 1) on Windows (your initial report mentioned "Artix Linux") and 2) you are launching nvim from the user home folder.

Oh no, I'm definitely not on windows. Yes I was launching nvim from the user home folder.

> So;
> 
>     1. On Windows, you will most likely have to enable the yet undocumented gotestsum runner by providing the `runner = "gotestsum"` option when setting up neotest-golang: [feat: gotestsum #150](https://github.com/fredrikaverpil/neotest-golang/pull/150)
> 
>     2. Please start `nvim` from the Go project folder instead (regardless of OS). A lot of issues you might have here (slowness being one) would most likely stem from this problem. Imagine the sheer number of files that Neotest might be parsing that resides under your user home folder.
Oh I somehow thought that nvim and neotest would just parse the directory of the currently opened file!
 Tried this yesterday and now I'm getting the same error as the other discussion I saw https://github.com/fredrikaverpil/neotest-golang/discussions/161 so I guess I'll try the solution mentioned there.

**Reply by fredrikaverpil** [reply]
> I'm definitely not on windows.

I somehow figured you were on Windows due to what your file tree looked like... sorry about that. 

> Oh I somehow thought that nvim and neotest would just parse the directory of the currently opened file!

No, this is not the case. Imagine starting neovim and immediately open up the test summary. I believe it starts processing (recursively) from your current working directory for _all_ adapters you have loaded.

A lot of Neovim plugins, like Telescope, operates from the assumption that you want to list or search files/folders recursively from the current working directory (`vim.fn.getcwd()`, where you start nvim from). There's this notion that you might want to change environment if you move from one project to another. A lot of workflows depend on this notion. So unless you actively change the cwd inside Neovim, you might see slowness across multiple plugins because of this.

> I'm getting the same error as the other discussion I saw https://github.com/fredrikaverpil/neotest-golang/discussions/161 so I guess I'll try the solution mentioned there.

Hm, so you suspect `go test` running through Neotest might also not output to stdout/stderr?
It would be super interesting to see what a full Neotest debug log would look like (instructions on how to enable that in the README) if you have the ability to produce that. It seems that on some platforms/distributions of Go, there output is not handled the same way (??). If you do this, please launch nvim from your project root (not user home) and run a single test. Then attach the full log here. This will show the exact command that gets executed along with output is captured from that execution. It might provide a lot of value in better understanding if I can e.g. add a warning that can accurately detect when this kind of problem is hit in case you are in experiencing the same issue as as described in #161.

You can also try out switching the test runner to `gotestsum` (you can find instructions in that thread or in #150). When using gotestsum, the adapter won't rely on that `go test` output is caught from stdout/stderr anymore.

As a final note, I noticed the `go.mod` file you provided says `module main.go`. For some reason, this is not a valid module name, or at least `go list` it cannot be used with `go test`, which the adapter will do if you attempt to run all tests in a file. If you change this into e.g. `module maingo`, `module main.yo`, `module yolo.go` or `module main.go/yolo` then it is valid. Not sure if this is a `go test` bug, a `go list` bug or something which is known and for some reason requires special treatment. To me it seems the exact problem is hit if you name the module `main.go` and nothing else and the result is you will see the "cannot be associated" warning in Neovim as a result. Note that this is only a problem when running all tests in a file.

You can run `go test -v main.go` and see how it doesn't run any tests. But if you change the module name in go.mod into `maingo` and then execute `go test -v maingo` you can see it executes the `TestHello` test. I think the reason for all this is that go might be confusing the package/module name (which is what we want to define) with an existing file name. I just added https://github.com/fredrikaverpil/neotest-golang/issues/165 and I might look into this... although it feels like an extreme edge case.

**Reply by Hobbyist11** [reply]
I've got it all working very well now! Using project.nvim and defining go.mod as a project root (for being in the cwd of the current project),
 somehow the issue: 

>Hm, so you suspect go test running through Neotest might also not output to stdout/stderr?
It would be super interesting to see what a full Neotest debug log would look like (instructions on how to enable that in the README) if you have the ability to produce that. It seems that on some platforms/distributions of Go, there output is not handled the same way (??). If you do this, please launch nvim from your project root (not user home) and run a single test. Then attach the full log here. This will show the exact command that gets executed along with output is captured from that execution. It might provide a lot of value in better understanding if I can e.g. add a warning that can accurately detect when this kind of problem is hit in case you are in experiencing the same issue as as described in https://github.com/fredrikaverpil/neotest-golang/discussions/161.

was fixed as well. Apologies if I was not able to perform the debug logs for it and try to also solve it myself. I got really busy. Anyways, can't thank you enough! I've got it all working now and it really is nice to just hit some keys and test files instead of having to open terminal on nvim and run go test!

**Reply by fredrikaverpil** [reply]
I'm glad you got it sorted and can enjoy writing tests in Go! 🥳
